### ISSUE TITLE
missing "tests" folder 

### ISSUE THREAD

**Issue by mastercad** [php-mock/php-mock#25]
Hi,

i have the Problem, that the Tests are missing if i install the version 2.1 via composer. If i look in your repository in github, it seems composer.json are the reason for that:

"archive": {
        "exclude": ["/tests"]
    }

Problem here is the Code in vendor/php-mock/php-mock/autoload.php where you create class alias for a class who does not exists in the project. I cant really use your package with this behaviour.

**Comment by mastercad** [comment]
i dont think so. this classes are test classes. so i think the tests should delivered with composer require. then the problem is away. if you wont to deliver tests in this project, i think i check is needed, if the tests folder exists. but i dont know future problems if the alias not created.

**Comment by webimpress** [comment]
@mastercad I can see the issue you've described, but I can't replicate it. I mean: always when I install package `php-mock-phpunit` via composer `php-mock` is also installed but I can see `tests/` directory in `vendor`. Even if I move packages to `require` section and do `composer update --no-dev` still I can see `tests/` directory. It looks like this `tests/` directory is only excluded on building packages.

What are you exactly doing that you got this error? Thanks!

**Comment by mastercad** [comment]
hmm. looks like is fixed without any solution? i have anymore required "composer require --dev php-mock/php-mock" and now the tests folder exists.

but i think its a good solution to check if the folder there in the autoloader file to prevent this issue in the past.

**Comment by meetmatt** [comment]
Reproducible with --prefer-dist flag (so that composer fetches the archive instead of cloning full repo from github)

**Comment by webimpress** [comment]
@meetmatt 
Can you provide more information, please?
I've tried it, so I've run:
```console
composer require --dev --prefer-dist php-mock/php-mock-phpunit
```

and I have installed these packages:
```
  - Installing php-mock/php-mock (2.1.1): Loading from cache
  - Installing php-mock/php-mock-integration (2.0.0): Loading from cache
  - Installing php-mock/php-mock-phpunit (2.3.0): Loading from cache
```

but all of them have `tests` directory in vendor, so still I cannot reproduce the issue.

**Comment by meetmatt** [comment]
@webimpress oh, sorry, I've mixed the "reproducible" and "fixed" commands.

Anyway, I will try to find the root cause, which is not in the missing tests folder, but with how the code is structured and autoloader configuration.

May I ask you, why have you decided to make this configuration:
```
    "autoload": {
        "psr-4": {
            "phpmock\\": ["classes/", "tests/"]
        },
        "files": ["autoload.php"]
    }
```

If tests folder contains only tests (of the library itself), and the autoloader hack is needed only for tests, then I believe it should look like this:
```
    "autoload": {
        "psr-4": {
            "phpmock\\": "classes/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "phpmock\\test\\": "tests/"
        },
        "files": ["autoload.php"]
    }
```
And all test files should be put under a \phpmock\test namespace.
Will test it out and let you know in a couple of days.

**Comment by meetmatt** [comment]
By the way, if you really want to prevent people from fetching tests folder during composer install, then besides composer.json:
```
    "archive": {
        "exclude": ["/tests"]
    }
```
You can also add this to `.gitattributes`:
```
tests/ export-ignore
```

But before that we need to close this issue first of course :)

Quote from SO:
> When you run the update or install command with --prefer-dist , Composer tries to download the archive on github. You can remove the test directory from the archives by putting this in a .gitattributes file in the root directory of your project: Tests/ export-ignore.

>

**Comment by webimpress** [comment]
@meetmatt 

> May I ask you, why have you decided
it was already like that before. I think it is because we are using tests in related components (like php-mock-phpunit), but these are actually tests.

The problem is with autoloader as it should work in related component tests.

I'm not the author of the library, just maintaining it now. The whole architecture is not perfect, and the question is how we can fix it easily without any breaking changes.

So, to summarize: I can replicate the issue when I remove manually the `tests/` folder, but using composer commands I can't. So I don't really know what is the real-life test case.

**Comment by meetmatt** [comment]
It's because of the composer cache.
Do `composer clear-cache` and then `composer require --dev --prefer-dist php-mock/php-mock-phpunit`.
You shall see `cloning` in the log.

**Comment by webimpress** [comment]
@meetmatt  I've tried it as well but still I do have `tests/` directories:

```console
$ composer require --dev --prefer-dist php-mock/php-mock-phpunit
Using version ^2.3 for php-mock/php-mock-phpunit
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 31 installs, 0 updates, 0 removals
  - Installing phpunit/php-text-template (1.2.1): Downloading (100%)         
  - Installing php-mock/php-mock (2.1.1): Downloading (100%)         
  - Installing php-mock/php-mock-integration (2.0.0): Downloading (100%)         
  - Installing sebastian/version (2.0.1): Downloading (100%)         
  - Installing sebastian/resource-operations (2.0.1): Downloading (100%)         
  - Installing sebastian/recursion-context (3.0.0): Downloading (100%)         
  - Installing sebastian/object-reflector (1.1.1): Downloading (100%)         
  - Installing sebastian/object-enumerator (3.0.3): Downloading (100%)         
  - Installing sebastian/exporter (3.1.0): Downloading (100%)         
  - Installing sebastian/environment (4.2.2): Downloading (100%)         
  - Installing sebastian/diff (3.0.2): Downloading (100%)         
  - Installing sebastian/comparator (3.0.2): Downloading (100%)         
  - Installing phpunit/php-timer (2.1.1): Downloading (100%)         
  - Installing phpunit/php-file-iterator (2.0.2): Downloading (100%)         
  - Installing theseer/tokenizer (1.1.2): Downloading (100%)         
  - Installing sebastian/code-unit-reverse-lookup (1.0.1): Downloading (100%)         
  - Installing phpunit/php-token-stream (3.0.1): Downloading (100%)         
  - Installing doctrine/instantiator (1.2.0): Downloading (100%)         
  - Installing symfony/polyfill-ctype (v1.11.0): Downloading (100%)         
  - Installing webmozart/assert (1.4.0): Downloading (100%)         
  - Installing phpdocumentor/reflection-common (1.0.1): Downloading (100%)         
  - Installing phpdocumentor/type-resolver (0.4.0): Downloading (100%)         
  - Installing phpdocumentor/reflection-docblock (4.3.1): Downloading (100%)         
  - Installing phpspec/prophecy (1.8.0): Downloading (100%)         
  - Installing phar-io/version (2.0.1): Downloading (100%)         
  - Installing phar-io/manifest (1.0.3): Downloading (100%)         
  - Installing myclabs/deep-copy (1.9.1): Downloading (100%)         
  - Installing sebastian/global-state (2.0.0): Downloading (100%)         
  - Installing phpunit/php-code-coverage (6.1.4): Downloading (100%)         
  - Installing phpunit/phpunit (7.5.12): Downloading (100%)         
  - Installing php-mock/php-mock-phpunit (2.3.0): Downloading (100%)         
sebastian/global-state suggests installing ext-uopz (*)
phpunit/phpunit suggests installing phpunit/php-invoker (^2.0)
Writing lock file
Generating autoload files
```

Composer 1.8.5, PHP 7.1

**Comment by meetmatt** [comment]
`Downloading (100%)` - it means that composer downloads archived version from packagist.org...
In my case I don't use packagist in favour of self-hosted satis mirror, probably that's the reason why I can reproduce it.
I can propose another way of reproducing it:
1. `composer clear-cache`
2. Add VCS repository URL to composer.json in the project where you want to reproduce the issue:
```
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/php-mock/php-mock"
        }
    ],
```
3. `composer require --dev --prefer-dist php-mock/php-mock`

That must work

**Comment by webimpress** [comment]
@meetmatt 

> That must work

Nope. Still the same result...  Anyway, I know what is the issue. I do not know what would be the best solution. What solution would you suggest? Do you think check if directory exists in the autoloader is sufficient for now?

At the end we would need to release new version, which support only PHP 7.2 and PHPUnit 8.0. Then all should be much easier. But in that occasion I would think what other changes we can make, that we don't need to use tests from php-mock in other libraries.

**Comment by meetmatt** [comment]
I would go with moving the autoloader to autoload-dev along with the tests folder like this:
```
"autoload": {
        "psr-4": {
            "phpmock\\": "classes/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "phpmock\\": "tests/"
        },
        "files": ["autoload.php"]
    }
```

**Comment by meetmatt** [comment]
I will be able to send you a PR with the mentioned solution by Saturday, June 1st

**Comment by webimpress** [comment]
@meetmatt This solution is not gonna work as then `php-mock-phpunit` using `tests/` and anything what is `autoload-dev` will be not available in package which consumes that one.

**Comment by meetmatt** [comment]
Started working on the PR and realized that the test case in php-mock uses PHPUnit to test itself.
For that purpose there's an AbstractMockTest which extends the PHPUnit TestCase.
Which is later required in the php-mock-phpunit as "production" code.
All of this is really screwed up and impossible to fix with existing code without circular dependencies.
I consider 2 options:
1. Copying the AbstractMockTest class in the php-mock-phpunit.
2. Moving the AbstractMockTest class to php-mock-phpunit and refactoring the local php-mock tests to not rely on it but rather on plain PHPUnit TestCase.

Will try first one first as it's safer.

**Comment by meetmatt** [comment]
So these 2 PR's should do the work, but I need more time to check if it works in integration environment:
1. https://github.com/php-mock/php-mock/pull/26
2. https://github.com/php-mock/php-mock-phpunit/pull/36

**Comment by webimpress** [comment]
@meetmatt  This is what I was talking about. This abstract test is used in related packages, I think in all of them -prophecy, -mockery, -phpunit.

> Copying the AbstractMockTest class in the php-mock-phpunit.

Not gonna work as we need it also in other packages, for example see:
https://github.com/php-mock/php-mock-prophecy/blob/master/tests/PHPProphetTest.php#L5

That's why it was done that way. And in general, maybe it's not perfect but it was working, and please remember - php-mock is in general dev dependency. So at the end it doesn't really matter. That's why I'd like to have the simplest fix ever, without rewriting the whole libraries. Rewrite we can plan for v3 :)

**Comment by meetmatt** [comment]
The simplest fix then would be to remove the exclude tests from composer.json 🤷🏻‍♂️

**Comment by webimpress** [comment]
@meetmatt 
> The simplest fix then would be to remove the exclude tests from composer.json
yeah, or add the check in autoload.php as @mastercad suggested.

**Comment by meetmatt** [comment]
Hm, then first of all we gonna have to move the “production” files from tests dir to classes dir. It’s not a breaking change since they share the same namespace (composer autoloader will figure out the new location), so can be included in the proposed solution. Then it’s just a matter of composer update for the dependent libraries.
